.circleci/config.yml

```diff
@@ -0,0 +1,139 @@
+defaults: &defaults
+  working_directory: ~/loga
+  steps:
+    - checkout
+    - &restore_gems
+      restore_cache:
+        keys:
+          - v2-gems-{{ checksum "Gemfile" }}-{{ checksum "Appraisals"}}-{{ .Environment.CIRCLE_JOB }}
+          - v2-gems-{{ checksum "Gemfile" }}
+    - &bundle_install
+      run: bundle install --jobs 4 --path=vendor/bundle --retry 3
+    - run: bundle exec appraisal install
+    - &cache_gems
+      save_cache:
+        key: v2-gems-{{ checksum "Gemfile" }}-{{ checksum "Appraisals"}}-{{ .Environment.CIRCLE_JOB }}
+        paths:
+          - ~/loga/gemfiles/vendor/bundle
+          - ~/loga/vendor/bundle
+    - attach_workspace:
+        at: ./tmp
+    - run: RACK_ENV=development bundle exec appraisal rspec
+    - run: RACK_ENV=production bundle exec appraisal rspec
+    - run: ./tmp/cc-test-reporter format-coverage -t simplecov -o $(ruby -e 'puts "tmp/codeclimate.#{RUBY_VERSION}.json"') coverage/.resultset.json
+    - persist_to_workspace:
+        root: tmp
+        paths:
+          - codeclimate.*.json
+    - store_artifacts:
+        path: coverage
+
+version: 2
+jobs:
+  build:
+    working_directory: ~/loga
+    docker:
+      - image: circleci/ruby:2.5.1
+    steps:
+      - run:
+          name: Download cc-test-reporter
+          command: |
+            mkdir -p tmp/
+            curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./tmp/cc-test-reporter
+            chmod +x ./tmp/cc-test-reporter
+      - persist_to_workspace:
+          root: ./tmp
+          paths:
+            - cc-test-reporter
+  upload-coverage:
+    working_directory: ~/loga
+    docker:
+      - image: circleci/ruby:2.5.1
+    steps:
+      - attach_workspace:
+          at: ./tmp
+      - run:
+          name: Upload coverage results to Code Climate
+          command: |
+            ./tmp/cc-test-reporter sum-coverage tmp/codeclimate.*.json -o tmp/codeclimate.total.json
+            ./tmp/cc-test-reporter upload-coverage -i tmp/codeclimate.total.json -r $CODECLIMATE_REPO_TOKEN
+  ruby-2.2:
+    docker:
+      - image: circleci/ruby:2.2.10
+    <<: *defaults
+  ruby-2.3:
+    docker:
+      - image: circleci/ruby:2.3.7
+    <<: *defaults
+  ruby-2.4:
+    docker:
+      - image: circleci/ruby:2.4.4
+    <<: *defaults
+  ruby-2.5:
+    docker:
+      - image: circleci/ruby:2.5.1
+    <<: *defaults
+  rubocop:
+    working_directory: ~/loga
+    docker:
+      - image: circleci/ruby:2.5.1
+    steps:
+      - checkout
+      - *restore_gems
+      - *bundle_install
+      - *cache_gems
+      - run: bundle exec rubocop
+  push-to-rubygems:
+    working_directory: ~/loga
+    docker:
+      - image: circleci/ruby:2.5.1
+    steps:
+      - checkout
+      - *restore_gems
+      - *bundle_install
+      - *cache_gems
+      - run:
+          name: Build gem
+          command: gem build $CIRCLE_PROJECT_REPONAME.gemspec
+      - run:
```

Comment:
🙋 This step _must_ be a `deploy` key. It's important that it is only attempted once.

Reply:
What is `deploy` key? I am lost on this one.

Reply:
@sgerrand I don't know that the deploy key provides that guarantee. According to the [documentation](https://circleci.com/docs/2.0/configuration-reference/#deploy) its only concerns jobs that run in parallel.

> In general deploy step behaves just like run with one exception - in a job with parallelism, the deploy step will only be executed by node #0 and only if all nodes succeed. Nodes other than #0 will skip this step.

Reply:
So it is in the doc.. Those search boxes at the top are pointless 🙈